### PR TITLE
[receiver/couchbase] Couchbase REST Client

### PR DESCRIPTION
**Pull request by cpheps**
**Description:**
- Added REST client for collecting metrics from the Couchbase API.
- Added Data models to represent API responses.
  - Choose this over doing a `map[string]interface{}` parsing as key response fields are static and there are a large number of nested objects
- Removed references to `couchDB` that were overlooked in last PR (Used the couchDB receiver as an example initially)

**Testing:**
Added unit tests for client and files of actual API responses to test against. Verified api responses do not contain sensitive data.

**Note**: Chose not to update changelog as this is further development on a receiver that is currently listed under the `New Components` section.

### CODE REVIEW

**Comment by cpheps** [comment]
@jpkrohling Per discussion on the last [PR concerning TLS](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7046#discussion_r781275624). I removed the line in the default config that created a blank `TLSClientSetting` as it's the zero value anyways.

As far as requiring TLS when user/pass is required I think for now we should not impose that. I agree that TLS should be enabled on systems but I'm not sure we'd want to require that if the technology itself does not require TLS in the base install. 

For example base installation for couchbase (detailed [here](https://docs.couchbase.com/server/current/install/ubuntu-debian-install.html#install-using-apt)) does not force TLS configuration. I believe the configuration of the receiver should match the base install of the technology. We don't want to impose security setup on the thing we monitoring, and if we require a TLS config (even if it's just to say we're operating insecure) that adds an additional barrier to speedy setup. 

I think as long as we have good documentation on how to do TLS configuration any users that choose to have their system configured as such should be able to set up the receiver.

**Comment by jpkrohling** [comment]
> We don't want to impose security setup on the thing we monitoring

That's reasonable. Should we at least then add a warning to the logs?

**Comment by cpheps** [comment]
> That's reasonable. Should we at least then add a warning to the logs?

I would say no. Since setting up TLS on Couchbase is a choice I don't think it's the collector's job to warn the user they are using a non-secure connection. 

In the case that the user does have TLS on Couchbase and misconfigures the collector it should fail to connect and produce an appropriate error log.

**Comment by jpkrohling** [comment]
Alright, sounds good to me.

File: receiver/couchbasereceiver/client_test.go [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+func TestNewClient(t *testing.T) {
+	testCase := []struct {
+		desc        string
+		cfg         *Config
+		host        component.Host
+		logger      *zap.Logger
+		expectError error
+	}{
+		{
+
```
Comment (djaglowski) [inline receiver/couchbasereceiver/client_test.go:51]:
```suggestion
```

File: receiver/couchbasereceiver/client_test.go [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+				require.Equal(t, tc.cfg.Endpoint, actualClient.hostEndpoint)
+				require.Equal(t, tc.logger, actualClient.logger)
+				require.NotNil(t, actualClient.client)
+			}
+		})
+	}
+}
+
+func TestGetClusterDetails(t *testing.T) {
+
```
Comment (djaglowski) [inline receiver/couchbasereceiver/client_test.go:105]:
```suggestion
```

File: receiver/couchbasereceiver/client_test.go [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, tc.testFunc)
+	}
+}
+
+func TestGetBuckets(t *testing.T) {
+
```
Comment (djaglowski) [inline receiver/couchbasereceiver/client_test.go:173]:
```suggestion
```

File: receiver/couchbasereceiver/client_test.go [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, tc.testFunc)
+	}
+}
+
+func TestGetBucketStats(t *testing.T) {
+
```
Comment (djaglowski) [inline receiver/couchbasereceiver/client_test.go:241]:
```suggestion
```

File: receiver/couchbasereceiver/client_test.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+	defer func() {
+		closeErr := file.Close()
+		require.NoError(t, closeErr)
+	}()
+
+	data, err := io.ReadAll(file)
+	require.NoError(t, err)
+
+	return data
+}
```
Comment (djaglowski) [inline receiver/couchbasereceiver/client_test.go:333]: We haven't been following a standard on this, but perhaps we should. `ioutil.ReadFile` is much more concise and handles the edge cases. Do you think it's worth standardizing on that so we don't have to worry about this level of detail in each scraper?

File: receiver/couchbasereceiver/testdata/apiresponses/get_clusters_response.json [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        },
+        "hdd": {
+            "total": 439074951168,
+            "quotaTotal": 439074951168,
+            "used": 30735246579,
+            "usedByData": 61994688,
+            "free": 408339704589
+        }
+    }
+}
```
Comment (djaglowski) [inline receiver/couchbasereceiver/testdata/apiresponses/get_clusters_response.json:351]:
```suggestion
}

```

File: receiver/couchbasereceiver/models.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package couchbasereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/couchbasereceiver"
+
+// Cluster response models
+
+type clusterResponse struct {
```
Comment (djaglowski) [inline receiver/couchbasereceiver/models.go:19]: The concern I always have with unmarshaling responses into structs is that we may lose the ability to differentiate between missing data points and actual zero values. Do you know if we can count on the couchbase response to always be complete?

Comment (djaglowski) [reply receiver/couchbasereceiver/models.go:19]: [This](https://docs.couchbase.com/server/current/rest-api/rest-bucket-stats.html#response-2) says `Depending on the level of bucket activity, there may be more detail for each field or less.`, which isn't perfectly clear, but certainly begs the question as to whether or not fields will be omitted due to lack of relevant activity.

File: receiver/couchbasereceiver/client_test.go (hunk 5 above)
Comment (cpheps) [reply receiver/couchbasereceiver/client_test.go:333]: I don't have strong feelings one way or another. Since this is just a test helper instead of production and a very controlled environment with what it's loading I'm not too concerned about edge cases but also I'm find with just using `ioutil.ReadFile` if that's what we're doing elsewhere in the code base. I'll update.

File: receiver/couchbasereceiver/testdata/apiresponses/get_clusters_response.json (hunk 6 above)
Comment (cpheps) [reply receiver/couchbasereceiver/testdata/apiresponses/get_clusters_response.json:351]: Saw this on a few of the JSON files I'll update.

File: receiver/couchbasereceiver/models.go (hunk 7 above)
Comment (cpheps) [reply receiver/couchbasereceiver/models.go:19]: If we switch to  pointers for data point values we won't lose that differentiation which is something I can do for the few metrics in the structs. The bulk of metrics will be in a `map[string]interface{}` on the `bucketStats` struct. 

That being said the ones I did specifically add fields for should be safe. I can add pointer fields for those just so we can null check them (would be the same as an `val, ok := map[key]` type check but no type casting of the value).

Comment (djaglowski) [reply receiver/couchbasereceiver/models.go:19]: Up to you - just need to make sure we are not generating phantom data points.

**Review by djaglowski — APPROVED** [review]
LGTM

File: receiver/couchbasereceiver/client.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		},
+		logger: logger,
+	}, nil
+}
+
+func (c *couchbaseClient) GetClusterDetails(ctx context.Context) (*clusterResponse, error) {
+	var clusterInfo clusterResponse
+
+	if err := c.get(ctx, clusterPath, &clusterInfo); err != nil {
+		c.logger.Error("Failed to retrieve cluster data", zap.Error(err))
```
Comment (jpkrohling) [inline receiver/couchbasereceiver/client.go:79]: This would generate duplicate entries in the log, wouldn't it? The call chain would eventually log the error, and this code here is logging it as well.

Comment (cpheps) [reply receiver/couchbasereceiver/client.go:79]: Yeah, I think you're right. Things that call `get` should likely not log the error. I think the message is useful to indicate what instance of the `get` call failed but likely not also logging the error.

If you agree I'll remove the `zap.Error` fields from the top level API calls.

Comment (jpkrohling) [reply receiver/couchbasereceiver/client.go:79]: I would remove all c.logger.Error calls that are followed by "return err". Or change them to debug, so that people will get more details only when the appropriate log-level is set.

Comment (cpheps) [reply receiver/couchbasereceiver/client.go:79]: I'll change them to debug, if nothing else if a bug gets filed against these it'll help us figure out what's going wrong.

Comment (cpheps) [reply receiver/couchbasereceiver/client.go]: @jpkrohling I changed all the logs in that file to debug logs. I think the error messages capture what is happening and the additional debug logging can help a user or us pinpoint what's going on.

**Review by jpkrohling — APPROVED** [review]
